Add New API Call To Get All Unique Fields For Given Entity

vinuvarshith opened the pull request:
Overview
----------------------------------------
This is a new api that will go through the indexes of an entity table and return all unique fields (including combination indexes).
This feature was proposed in a disscusion here
https://github.com/eileenmcnaughton/nz.co.fuzion.csvimport/issues/13

Before
----------------------------------------
It's not currently possible to get a list of unique fields (or combinations) for a given entity.

After
----------------------------------------
A new api is added to get a list of all unique fields for a given entity.

Technical Details
----------------------------------------
Sample API call using 'Contribution' entity:
`$result = civicrm_api3('Contribution', 'getunique');`

Result:
```
{
    "is_error": 0,
    "version": 3,
    "count": 2,
    "values": {
        "UI_contrib_trxn_id": [
            "trxn_id"
        ],
        "UI_contrib_invoice_id": [
            "invoice_id"
        ]
    }
}
```

civicrm-builder commented:
Can one of the admins verify this patch?

MiyaNoctem commented on api/v3/Generic/Getunique.php:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @return mixed
+ */
+function civicrm_api3_generic_getUnique($apiRequest) {
```
You should probably add a _spec function for this API action too. Check:

https://docs.civicrm.org/dev/en/latest/framework/api-architecture/#spec

eileenmcnaughton replied:
std is lcase - ie getunique

eileenmcnaughton replied:
yeah - this is the api functions std not something general

colemanw commented:
If this is a greenfield project, consider using Api V4. https://github.com/civicrm/api4

eileenmcnaughton commented:
@colemanw this is proposed to build on extensions that work on api v3 already & TBH we already rejected a change for core & this is the less intrusive core change proposed coming off that. I'm OK with this in principle. Note if we DID add the unique custom field type this should also be extended to return those fields 

@vinuvarshith you don't need to query the DB - the DAOs have an indices function you can use. 

Note this will need a unit test - also the PR description should include a code snippet calling this api & what the response looks like.

vinuvarshith commented:
@eileenmcnaughton 
I have updated code to use indices functions in DAOs instead of an SQL query.
I have added 2 testcases one for testing 'getunique' (for 'Contact' and for 'Contribution' entities)

eileenmcnaughton commented:
@vinuvarshith if this is ready for review can you remove the WIP tag. Also check the comment about casing - I know it's a bit ugly to do getunique - but consistent with getfields getsingle etc

seanmadsen commented:
I'd like to request the following change to satisfy [r-doc](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc):

* Can you add a corresponding PR against [dev-docs](https://github.com/civicrm/civicrm-dev-docs) which inserts a new section to the [API Actions](https://docs.civicrm.org/dev/en/latest/api/actions/) page describing the action?

vinuvarshith commented:
@seanmadsen 
Added PR against dev-docs here
https://github.com/civicrm/civicrm-dev-docs/pull/522

eileenmcnaughton commented:
* ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) __Pass__
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __Pass__
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __Pass__
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) __Passd__
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __Pass__
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __Pass __: <!-- Describe -->
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) __Pass__: <!-- Describe -->
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __Pass__: <!-- Describe -->

totten commented:
Small correction -- it looks like the tests weren't properly run on this PR. (Jenkins wanted confirmation from an admin before it would run the tests.)

I suspect there's no real harm in this case, though I had a small issue where `civilint` failed on some unrelated changes. It's updated as part of #12029 (2a7d2b9c9b30915ab63fa3cb622f746a001ed56d).
